### DISCUSSION TITLE
Proposal: Statement lambda presevation

### DISCUSSION THREAD

**Discussion by biqas** [dotnet/csharplang#1616] [Language Ideas] [answered]
Hello,

I would like to make a proposal regarding to Statement lambdas.

Currently we have following situation:

```csharp
private void X()
{
	Y(x =>
	{
		var z = x + 42;

		return z.ToString();
	});
}

private void Y(Func<int, string> someFn)
{
   // here we just get a delegate
}
```

The proposal is to preserve the statement lambda so that we can access the full LamdaExpresseionSyntax:

```csharp
private void X()
{
	Y(syntax x =>
	{
		var z = x + 42;

		return z.ToString();
	});
}

private void Y(syntax Func<int, string> someFn)
{
   // here we should get the delegate and LamdaExpresseionSyntax
   // If we want to access the delegate as usual, there would be no differences.
   someFn(42);
   
   // If we want to access the syntax, one proposal could be
   // (no ambitions about used keywords or symbols)
   LamdaExpresseionSyntax someFnSyntax = syntax:someFn;
   var block = someFnSyntax.Block;
   // some transformations for line var z = x + 42 to var z = x + 100;
   … NumericLiteralToken.Value = 100;
   // compile the new expression to a delegate
   someFn:compile
}
```

So what benefits you would have:
- new kind of reflection based on syntax trees. (I guess this kind of reflection needs also no maintenance, because what ever syntax trees provide you can use.)
- code generators could use statements from caller side and transform them

The idea/proposal can be applied to more then delegates only, but that is maybe a different story.

**Comment by stakx** [comment]
Do you know about LINQ expression trees (`System.Linq.Expressions`)? They have some increasingly serious limitations (see e.g. #158) but it would seem more reasonable to update an existing mechanism, instead of introducing a new one having a large functional overlap.

**Comment by biqas** [comment]
@stakx Hi I know very well about `System.Linq.Expressions`, the expressions work differently. I'm not talking about creating a synthetic syntax tree. The main idea is to preserve the exact syntax you wrote and if you want to pass the whole syntax to callee just mark them with some keyword.

And I disagree with introducing functional overlap, because this is already there. Both `System.Linq.Expressions` and `Microsoft.CodeAnalysis.CSharp.Syntax.SyntaxTree`.

**Comment by CyrusNajmabadi** [comment]
Take a look at the [CallerArgumentExpression](https://github.com/dotnet/csharplang/blob/master/proposals/caller-argument-expression.md) proposal.  It would give you what you wanted.  You would be able to get the source-code for the expression passed to you.  You could then use Roslyn to parse that source if you wantd to get the LambdaExpressionSyntax (presuming, of course, that you were actually passed a lambda).

**Comment by biqas** [comment]
@CyrusNajmabadi the proposal seems promising. Will follow and see what the design meetings results will be.

**Comment by svick** [comment]
So, what's the big advantage that having access to Roslyn syntax tree would give you here, compared with improved support for LINQ expression trees?

**Comment by biqas** [comment]
@svick I guess the main benefits are for libraries which are generating code. so they can analyze passed syntax as it was written. Of course those generators currently does not exists I'm assuming if the feature already exists.

Serialization Generators
Pattern Generators
..

**Comment by svick** [comment]
@biqas 

I think you still haven't really answered the question. What *specific* advantage does having access to "syntax as it was written" give you? Could you give a concrete example of code that you could write with this proposal, but not with improved LINQ expression trees?

**Comment by biqas** [comment]
@svick 

Yes you are right I did not really answered your question :)

Ok let me try, because I'm not sure what you exactly mean by LINQ expression trees. The reason is based on the specs: https://www.ecma-international.org/publications/files/ECMA-ST/Ecma-334.pdf

`
Standard ECMA-334 5th Edition
Chapter: 11.7.3
A lambda expression F is compatible with an expression tree type Expression<D> if F is compatible with 
the delegate type D. This does not apply to anonymous methods, only lambda expressions. 

Certain lambda expressions cannot be converted to expression tree types: Even though the conversion 
exists, it fails at compile-time. This is the case if the lambda expression:
• Has a block body
• Contains simple or compound assignment operators
• Contains a dynamically bound expression
• Is async 

Anonymous functions may influence overload resolution, and participate in type inference. See §12.6 for further details.
`

`Statement lambdas, like anonymous methods, cannot be used to create expression trees` and I doubt that even if you extend/improve or enhance LINQ expression trees it would be possible. (As an example any new language feature then must be then included in expression trees and there are also other limitation how expressions are designed).

So the example given above you can't do because you cannot capture an statement lambda and transform it into an expression tree.

Advantage would be that it would be then possible :)

**Comment by svick** [comment]
@biqas 

So, if you want to support statement lambdas, you have two options:

1. Add support of statement lambdas to LINQ expression trees (see https://github.com/dotnet/csharplang/issues/158). This is certainly possible and should be relatively easy to do, especially since all the necessary types already exist in `System.Linq.Expressions`.
2. Create a whole new feature that means the compiler has to generate code that creates a different kind of trees. Even if those are the same trees that the compiler already uses, that's still a lot of work.

What makes option 2 better than option 1?

And even for features that don't currently have corresponding types in `System.Linq.Expressions`, improving expression trees still sounds like a significantly better option to me.

**Comment by svick** [comment]
Also, I don't quite understand the way your proposal suggests manipulating `syntax` is supposed to work.

First, Roslyn syntax trees are immutable, so you can't do something like change the `Value` of the `SyntaxToken` for a `LiteralExpressionSyntax`.

Second, I'm confused about syntax like `syntax:someFn` or `someFn:compile`. Since it seems the `syntax` parameter has some hidden data (like values of variables from the outer function used by the lambda), does it mean there is no way to pass that hidden data around? So e.g. if I wanted to write a helper library that manipulates and compiles syntax trees produced by `syntax`, I wouldn't be able to do that?

I think a much more reasonable way to do this would be to create a type that holds all the required information. E.g.:

```c#
private void Y(Syntax<Func<int, string>> someFn)
{
    someFn.Delegate.Invoke(42);
    
    LambdaExpressionSyntax someFnSyntax = someFn.Expression;

    var literalToken = someFnSyntax.DescendantNodes().OfType<LiteralExpressionSyntax>().Single().Token;

    // some transformations for line var z = x + 42 to var z = x + 100;
    var transformedSyntax = someFnSyntax.ReplaceToken(literalToken, SyntaxFactory.Literal(100));

    // compile the new expression to a delegate
    Func<int, string> newDelegate = someFn.Recompile(transformedSyntax);
}
```

**Comment by gafter** [comment, marked as answer]
The Linq expression tree APIs have support for statements. We just haven't implemented it yet. We discussed, recently in the LDM, extending support to all the current expression and statement forms. We have a volunteer who has already done preliminary implementation work. We're still discussing what we want to do.
